src/Persistence_matrix/include/gudhi/boundary_matrix/base_matrix_0000.h

```diff
@@ -0,0 +1,306 @@
+ *
+ *    Copyright (C) 2022 Inria
+ *
+ *    Modification(s):
+ *      - YYYY/MM Author: Description of the modification
+ */
+
+#ifndef BASE_MATRIX_0000_H
+#define BASE_MATRIX_0000_H
+
+#include <iostream>
+#include <vector>
+
+#include "../utilities/utilities.h"
+
+namespace Gudhi {
+namespace persistence_matrix {
+
+template<class Master_matrix>
+class Base_matrix
+		: public Master_matrix::Base_swap_option,
+		  public Master_matrix::Base_pairing_option
+{
+public:
+	using Column_type = typename Master_matrix::Column_type;
+	using boundary_type = typename Master_matrix::boundary_type;
+	using Row_type = void;
+
+	Base_matrix();
+	template<class Boundary_type = boundary_type>
+	Base_matrix(std::vector<Boundary_type>& orderedBoundaries);
+	Base_matrix(unsigned int numberOfColumns);
+	Base_matrix(const Base_matrix& matrixToCopy);
+	Base_matrix(Base_matrix&& other) noexcept;
+
+	template<class Boundary_type = boundary_type>
+	void insert_boundary(Boundary_type& boundary);
+	Column_type& get_column(index columnIndex);
+	Row_type get_row(index rowIndex);
+	void erase_last();
+
+	dimension_type get_max_dimension() const;
+	unsigned int get_number_of_columns() const;
+
+	dimension_type get_column_dimension(index columnIndex) const;
+
+	void add_to(index sourceColumnIndex, index targetColumnIndex);
+
+	void zero_cell(index columnIndex, index rowIndex);
+	void zero_column(index columnIndex);
+	bool is_zero_cell(index columnIndex, index rowIndex) const;
+	bool is_zero_column(index columnIndex);
+
+	index get_column_with_pivot(index simplexIndex);
+	index get_pivot(index columnIndex);
+
+	Base_matrix& operator=(Base_matrix other);
+	template<class Friend_master_matrix>
+	friend void swap(Base_matrix<Friend_master_matrix>& matrix1,
+					 Base_matrix<Friend_master_matrix>& matrix2);
+
+	void print();  //for debug
+
+private:
+	using swap_opt = typename Master_matrix::Base_swap_option;
+	using pair_opt = typename Master_matrix::Base_pairing_option;
+
+	typename Master_matrix::column_container_type matrix_;
+	dimension_type maxDim_;
+	index nextInsertIndex_;
+};
+
+template<class Master_matrix>
+inline Base_matrix<Master_matrix>::Base_matrix()
+	: Master_matrix::Base_swap_option(matrix_),
+	  Master_matrix::Base_pairing_option(matrix_, maxDim_),
+	  maxDim_(-1),
+	  nextInsertIndex_(0)
+{}
+
+template<class Master_matrix>
+template<class Boundary_type>
+inline Base_matrix<Master_matrix>::Base_matrix(std::vector<Boundary_type> &orderedBoundaries)
+	: Master_matrix::Base_swap_option(matrix_, orderedBoundaries.size()),
+	  Master_matrix::Base_pairing_option(matrix_, maxDim_),
+	  matrix_(orderedBoundaries.size()),
+	  maxDim_(0),
+	  nextInsertIndex_(orderedBoundaries.size())
+{
+	if constexpr (swap_opt::isActive_){
+		swap_opt::indexToRow_.resize(orderedBoundaries.size());
+		swap_opt::rowToIndex_.resize(orderedBoundaries.size());
+	}
+
+	for (unsigned int i = 0; i < orderedBoundaries.size(); i++){
+		Boundary_type& b = orderedBoundaries.at(i);
```

Comment:
I like to use at() as long as I am in testing mode. But I will replace them all at the end.

Reply:
With libstdc++ (gcc), you can define [`_GLIBCXX_ASSERTIONS`](https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_macros.html), or with libc++ (llvm) you can define [`_LIBCPP_DEBUG`](https://releases.llvm.org/14.0.0/projects/libcxx/docs/DesignDocs/DebugMode.html#basic-checks-libcpp-debug-0) or [`_LIBCPP_ENABLE_ASSERTIONS`](https://libcxx.llvm.org/UsingLibcxx.html#enabling-the-safe-libc-mode) depending on the version. These options make it so that `operator[]` aborts if called with an out-of-bounds argument.

Reply:
Ok, that will make things easier!